File: agents/utils/logger.py
"""
Structured logging framework for Python agents

Provides environment-aware logging with proper levels and context.
Replaces scattered print() statements with structured logging.
"""

import logging
import os
import sys
import json
from typing import Any, Dict, Optional
from datetime import datetime


class StructuredLogger:
    """
    Structured logger with JSON output and context support.
    
    Features:
    - Environment-aware log levels
    - Structured JSON output in production
    - Context injection (agent_id, user_wallet, etc.)
    - Performance tracking
    """
    
    def __init__(self, name: str, context: Optional[Dict[str, str]] = None):
        """
        Initialize structured logger.
        
        Args:
            name: Logger name (usually module name or agent ID)
            context: Default context to include in all logs
        """
        self.name = name
        self.context = context or {}
        self.logger = logging.getLogger(name)
        
        # Set log level from environment
        log_level = os.getenv("LOG_LEVEL", "INFO").upper()
        self.logger.setLevel(getattr(logging, log_level, logging.INFO))
        
        # Configure handler
        if not self.logger.handlers:
            handler = logging.StreamHandler(sys.stdout)
            
            # Use JSON formatter in production, simple formatter in development
            env = os.getenv("NODE_ENV", "development")
            if env == "production":
                handler.setFormatter(JSONFormatter())
            else:
                handler.setFormatter(logging.Formatter(
                    '%(asctime)s - %(name)s - %(levelname)s - %(message)s'
                ))
            
            self.logger.addHandler(handler)
    
    def _log(self, level: str, message: str, **kwargs):
        """Internal logging method with context injection."""
        # Merge default context with log-specific context
        log_context = {**self.context, **kwargs}
        
        # Get the logging method (debug, info, warning, error)
        log_method = getattr(self.logger, level.lower())
        
        # Log with extra context
        log_method(message, extra=log_context)
    
    def debug(self, message: str, **kwargs):
        """Log debug message."""
        self._log("DEBUG", message, **kwargs)
    
    def info(self, message: str, **kwargs):
        """Log info message."""
        self._log("INFO", message, **kwargs)
    
    def warning(self, message: str, **kwargs):
        """Log warning message."""
        self._log("WARNING", message, **kwargs)
    
    def error(self, message: str, **kwargs):
        """Log error message."""
        self._log("ERROR", message, **kwargs)
    
    def critical(self, message: str, **kwargs):
        """Log critical message."""
        self._log("CRITICAL", message, **kwargs)


class JSONFormatter(logging.Formatter):
    """JSON formatter for structured logging in production."""
    
    def format(self, record: logging.LogRecord) -> str:
        """Format log record as JSON."""
        log_data = {
            "timestamp": datetime.utcnow().isoformat(),
            "level": record.levelname,
            "logger": record.name,
            "message": record.getMessage(),
        }
        
        # Add any extra context from the record
        if hasattr(record, '__dict__'):
            for key, value in record.__dict__.items():
                if key not in ['name', 'msg', 'args', 'created', 'filename', 'funcName',
                              'levelname', 'levelno', 'lineno', 'module', 'msecs',
                              'message', 'pathname', 'process', 'processName',
                              'relativeCreated', 'thread', 'threadName', 'exc_info',
                              'exc_text', 'stack_info']:
                    log_data[key] = value
        
        return json.dumps(log_data)


# Global logger instance
_default_logger: Optional[StructuredLogger] = None


def get_logger(name: Optional[str] = None, **context) -> StructuredLogger:
    """
    Get or create a structured logger.
    
    Args:
        name: Logger name (defaults to calling module)
        **context: Default context for this logger
    
    Returns:
        StructuredLogger instance
    """
    global _default_logger
    
    if name is None:
        # Use calling module name
        import inspect
        frame = inspect.currentframe()
        if frame and frame.f_back:
            name = frame.f_back.f_globals.get('__name__', 'unknown')
    
    return StructuredLogger(name, context)


def configure_root_logger(level: str = "INFO"):
    """
    Configure root logger for the entire application.
    
    Args:
        level: Log level (DEBUG, INFO, WARNING, ERROR, CRITICAL)
    """
    logging.basicConfig(
        level=getattr(logging, level.upper(), logging.INFO),
        format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
    )

File: website/src/lib/logger.ts
/**
 * Structured logging framework for TypeScript/Next.js
 * 
 * Provides environment-aware logging with proper levels and context.
 * Replaces scattered console.log statements with structured logging.
 */

export enum LogLevel {
  DEBUG = 0,
  INFO = 1,
  WARN = 2,
  ERROR = 3,
  CRITICAL = 4,
}

export interface LogContext {
  [key: string]: any;
}

export interface LogEntry {
  timestamp: string;
  level: string;
  logger: string;
  message: string;
  context?: LogContext;
  error?: any;
}

class StructuredLogger {
  private name: string;
  private context: LogContext;
  private minLevel: LogLevel;

  constructor(name: string, context: LogContext = {}) {
    this.name = name;
    this.context = context;
    
    // Set log level from environment
    const envLevel = process.env.LOG_LEVEL?.toUpperCase() || 'INFO';
    this.minLevel = LogLevel[envLevel as keyof typeof LogLevel] ?? LogLevel.INFO;
  }

  private shouldLog(level: LogLevel): boolean {
    return level >= this.minLevel;
  }

  private formatLog(level: string, message: string, context?: LogContext, error?: any): LogEntry {
    return {
      timestamp: new Date().toISOString(),
      level,
      logger: this.name,
      message,
      context: { ...this.context, ...context },
      error: error ? this.serializeError(error) : undefined,
    };
  }

  private serializeError(error: any): any {
    if (error instanceof Error) {
      return {
        name: error.name,
        message: error.message,
        stack: error.stack,
        ...(error as any), // Include any additional properties
      };
    }
    return error;
  }

  private write(logEntry: LogEntry): void {
    // In production, output JSON for log aggregation
    if (process.env.NODE_ENV === 'production') {
      console.log(JSON.stringify(logEntry));
    } else {
      // In development, use readable format
      const contextStr = logEntry.context && Object.keys(logEntry.context).length > 0
        ? ` ${JSON.stringify(logEntry.context)}`
        : '';
      const errorStr = logEntry.error
        ? `\n${logEntry.error.stack || JSON.stringify(logEntry.error)}`
        : '';
      console.log(`[${logEntry.level}] ${logEntry.logger}: ${logEntry.message}${contextStr}${errorStr}`);
    }
  }

  debug(message: string, context?: LogContext): void {
    if (this.shouldLog(LogLevel.DEBUG)) {
      this.write(this.formatLog('DEBUG', message, context));
    }
  }

  info(message: string, context?: LogContext): void {
    if (this.shouldLog(LogLevel.INFO)) {
      this.write(this.formatLog('INFO', message, context));
    }
  }

  warn(message: string, context?: LogContext): void {
    if (this.shouldLog(LogLevel.WARN)) {
      this.write(this.formatLog('WARN', message, context));
    }
  }

  error(message: string, error?: any, context?: LogContext): void {
    if (this.shouldLog(LogLevel.ERROR)) {
      this.write(this.formatLog('ERROR', message, context, error));
    }
  }

  critical(message: string, error?: any, context?: LogContext): void {
    if (this.shouldLog(LogLevel.CRITICAL)) {
      this.write(this.formatLog('CRITICAL', message, context, error));
    }
  }
}

// Logger factory
const loggers = new Map<string, StructuredLogger>();

export function getLogger(name: string, context: LogContext = {}): StructuredLogger {
  const key = `${name}:${JSON.stringify(context)}`;
  
  if (!loggers.has(key)) {
    loggers.set(key, new StructuredLogger(name, context));
  }
  
  return loggers.get(key)!;
}

// Convenience function for creating loggers with request context
export function createRequestLogger(name: string, requestId?: string, userId?: string): StructuredLogger {
  return getLogger(name, {
    requestId,
    userId,
  });
}

// Default logger for quick usage
export const logger = getLogger('app');

